haystack/preview/components/generators/openai/_helpers.py

```diff
@@ -9,6 +22,151 @@
+    "gpt-3": 4096,  # Ref: https://platform.openai.com/docs/models/gpt-3
+    "gpt-4-32k": 32768,  # Ref: https://platform.openai.com/docs/models/gpt-4
+    "gpt-4": 8192,  # Ref: https://platform.openai.com/docs/models/gpt-4
+}
+
+
+#: Retry on OpenAI errors
+openai_retry = tenacity.retry(
+    reraise=True,
+    retry=tenacity.retry_if_exception_type(OpenAIError)
+    and tenacity.retry_if_not_exception_type(OpenAIUnauthorizedError),
+    wait=tenacity.wait_exponential(multiplier=OPENAI_BACKOFF),
+    stop=tenacity.stop_after_attempt(OPENAI_MAX_RETRIES),
+)
+
+
+def default_streaming_callback(token: str, **kwargs):
+    """
+    Default callback function for streaming responses from OpenAI API.
+    Prints the tokens to stdout as soon as they are received and returns them.
+    """
+    print(token, flush=True, end="")
+    return token
+
+
+@openai_retry
+def query_chat_model(url: str, headers: Dict[str, str], payload: Dict[str, Any]) -> List[str]:
+    """
+    Query ChatGPT without streaming the response.
+
+    :param url: The URL to query.
+    :param headers: The headers to send with the request.
+    :param payload: The payload to send with the request.
+    :return: A list of strings containing the response from the OpenAI API.
+    """
+    response = requests.post(url, headers=headers, data=json.dumps(payload), timeout=OPENAI_TIMEOUT)
+    raise_for_status(response=response)
+    json_response = json.loads(response.text)
+    check_truncated_answers(result=json_response, payload=payload)
+    check_filtered_answers(result=json_response, payload=payload)
+    return [choice["message"]["content"].strip() for choice in json_response["choices"]]
+
+
+@openai_retry
+def query_chat_model_stream(
+    url: str, headers: Dict[str, str], payload: Dict[str, Any], callback: Callable, marker: str
+) -> List[str]:
+    """
+    Query ChatGPT and streams the response. Once the stream finishes, returns a list of strings just like
+    self._query_llm()
+
+    :param url: The URL to query.
+    :param headers: The headers to send with the request.
+    :param payload: The payload to send with the request.
+    :param callback: A callback function that is called when a new token is received from the stream.
+        The callback function should accept two parameters: the token received from the stream and **kwargs.
+        The callback function should return the token that will be returned at the end of the streaming.
+    :param marker: A marker that indicates the end of the stream. It is used to determine when to stop streaming.
+    :return: A list of strings containing the response from the OpenAI API.
+    """
+    response = requests.post(url, headers=headers, data=json.dumps(payload), timeout=OPENAI_TIMEOUT, stream=True)
+    raise_for_status(response=response)
+
+    client = sseclient.SSEClient(response)
+    tokens = []
+    try:
+        for event in client.events():
+            if event.data == marker:
```

Comment:
I don't think OpenAI uses marker other than [DONE], please check

Reply:
You're right, seems like they don't: https://platform.openai.com/docs/api-reference/chat/create#stream I'm making it a constant then